internal/datastore/spanner/readonly.go

```diff
@@ -0,0 +1,219 @@
+
+		var tuples []*core.RelationTuple
+
+		if err := iter.Do(func(row *spanner.Row) error {
+			nextTuple := &core.RelationTuple{
+				ObjectAndRelation: &core.ObjectAndRelation{},
+				User: &core.User{
+					UserOneof: &core.User_Userset{
+						Userset: &core.ObjectAndRelation{},
+					},
+				},
+			}
+			userset := nextTuple.User.GetUserset()
+			err := row.Columns(
+				&nextTuple.ObjectAndRelation.Namespace,
+				&nextTuple.ObjectAndRelation.ObjectId,
+				&nextTuple.ObjectAndRelation.Relation,
+				&userset.Namespace,
+				&userset.ObjectId,
+				&userset.Relation,
+			)
+			if err != nil {
+				return err
+			}
+
+			tuples = append(tuples, nextTuple)
+
+			return nil
+		}); err != nil {
+			return nil, err
+		}
+
+		return tuples, nil
+	}
+}
+
+func (sr spannerReader) ReadNamespace(ctx context.Context, nsName string) (*core.NamespaceDefinition, datastore.Revision, error) {
+	ctx, span := tracer.Start(ctx, "ReadNamespace")
+	defer span.End()
+
+	nsKey := spanner.Key{nsName}
+	row, err := sr.txSource().ReadRow(
+		ctx,
+		tableNamespace,
+		nsKey,
+		[]string{colNamespaceConfig, colNamespaceTS},
+	)
+	if err != nil {
+		if spanner.ErrCode(err) == codes.NotFound {
+			return nil, datastore.NoRevision, datastore.NewNamespaceNotFoundErr(nsName)
+		}
+		return nil, datastore.NoRevision, fmt.Errorf(errUnableToReadConfig, err)
+	}
+
+	var serialized []byte
+	var updated time.Time
+	if err := row.Columns(&serialized, &updated); err != nil {
+		return nil, datastore.NoRevision, fmt.Errorf(errUnableToReadConfig, err)
+	}
+
+	ns := &core.NamespaceDefinition{}
+	if err := proto.Unmarshal(serialized, ns); err != nil {
+		return nil, datastore.NoRevision, fmt.Errorf(errUnableToReadConfig, err)
+	}
+
+	return ns, revisionFromTimestamp(updated), nil
+}
+
+func (sr spannerReader) ListNamespaces(ctx context.Context) ([]*core.NamespaceDefinition, error) {
+	ctx, span := tracer.Start(ctx, "ListNamespaces")
+	defer span.End()
+
+	iter := sr.txSource().Read(
+		ctx,
+		tableNamespace,
+		spanner.AllKeys(),
+		[]string{colNamespaceConfig},
+	)
+
+	allNamespaces, err := readAllNamespaces(iter)
+	if err != nil {
+		return nil, fmt.Errorf(errUnableToListNamespaces, err)
+	}
+
+	return allNamespaces, nil
+}
+
+func readAllNamespaces(iter *spanner.RowIterator) ([]*core.NamespaceDefinition, error) {
+	var allNamespaces []*core.NamespaceDefinition
+	if err := iter.Do(func(row *spanner.Row) error {
+		var serialized []byte
+		if err := row.Columns(&serialized); err != nil {
+			return err
+		}
+
+		ns := &core.NamespaceDefinition{}
+		if err := proto.Unmarshal(serialized, ns); err != nil {
+			return err
+		}
+
+		allNamespaces = append(allNamespaces, ns)
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return allNamespaces, nil
+}
+
+var queryTuples = sql.Select(
+	colNamespace,
+	colObjectID,
+	colRelation,
+	colUsersetNamespace,
+	colUsersetObjectID,
+	colUsersetRelation,
+).From(tableRelationship)
+
+var schema = common.SchemaInformation{
```

Comment:
Why is this in `readonly.go`?

Reply:
Because it's only used by the schema query filterer in `readonly.go`.